### ISSUE TITLE
Implement select a la Go

### ISSUE THREAD

**Issue by no-longer-on-githu-b** [GaloisInc/BoundedChan#4]
`select` would be immensely useful. It takes a sequence of (send/receive, action) pairs and executes the action belonging to the first send/receive that succeeds without affecting the other channels.

The interface could look like the following:

```
data Case b
    = forall a. Recv (BoundedChan a) (a -> IO b)
    | forall a. Send (BoundedChan a) a (IO b)

select :: [Case b] -> IO b
```

Usage:

```
loop :: Int -> BoundedChan Int -> BoundedChan Int -> BoundedChan () -> IO ()
loop n add get quit =
    select [ Recv add (\m -> loop (n + m) add quit)
           , Send get n (loop n add quit)
           , Recv quit (\_ -> return ())
           ]
```

I don't know how to implement it, but I heard it's an immense pain to do so.

See http://golang.org/ref/spec#Select_statements

**Comment by acw** [comment]
You're right in both cases: they're really handy, and they're a pain to implement.

I'm going to leave this open as a feature request, and would LOVE LOVE LOVE if someone out there in the open source community wanted to tackle this one. Otherwise, it's going to have to wait quite awhile until this sneaks back into the top of my TODO list.

**Comment by cstrahan** [comment]
So, I took a stab at writing a bounded channel library with the same behavior as Go's channels, and this is what what I came up with in terms of API (thus far -- I wrote most of this last weekend):

``` haskell
data Chan a

data Result a
    = Msg a
    | Closed

data Case a
    = forall b. Recv !(Chan b)
                     !(Result b -> IO a)
    | forall b. Send !(Chan b)
                     !b
                     !(IO a)

mkChan :: Int -> IO (Chan a)

closeChan :: Chan a -> IO ()

chanRecv :: Chan a -> IO (Result a)

chanRecvAsync :: Chan a -> IO (Maybe (Result a))

chanSend :: Chan a -> a -> IO ()

chanSendAsync :: Chan a -> a -> IO Bool

select :: [Case a] -> Maybe (IO a) -> IO a
```

The code is currently over here: https://github.com/cstrahan/bchan

I suppose I _could_ try to fold that into `BoundedChan`, but I think there are enough differences that it probably makes sense for this to be a separate library.

I'm not entirely sure what to call it (I originally named it `bchan`, but now I'm thinking of calling it `gchan` to pay homage to Go -- but at that point, maybe I should just call it `gochan`?). Also, I'd like to hear what people think of the API/types.

Once that's done, I figure I'll cut a release.

Thoughts?

**Comment by cstrahan** [comment]
Ooh, I think I already see something I'd like to change; this:

``` haskell
mkChan :: Int -> IO (Chan a)

closeChan :: Chan a -> IO ()
```

should perhaps be:

``` haskell
chanMake :: Int -> IO (Chan a)

chanClose :: Chan a -> IO ()
```

Just to keep those names consistent with the others. Or something more like that, any way.

EDIT: and `s/select/chanSelect/`, perhaps.

**Comment by acw** [comment]
Those look mostly reasonable to me. I'm a little torn, actually. I like this interface, and so integrating something like this into BoundedChan doesn't seem totally out of line. But I do like `chanRead` throwing an exception sometimes, rather than returning `Result`. Hmmm. If you're interested, maybe we could include your version, and then create an exception layer on top, too?

The one quibble I have is with your `Async` versions. By the type signatures, they sound like `try` rather than `async`. I.e., either this will work or will it fail and nothing will happen. As opposed to `async`, which I interpret as "this _will_ complete, at some point, probably later." They're both useful, but have different types, a la:

```
chanRead :: Chan a -> IO (Result a)
chanTryRead :: Chan a -> IO (Maybe (Result a))
chanAsyncRead :: Chan a -> IO (MVar (Result a))
```

Does that make sense?

Also, out of curiosity: what are the semantics of Go's `select`? Does it guarantee fairness across executions? Randomly select? Select the first one?

**Comment by no-longer-on-githu-b** [comment]
@acw Linear random distribution. Uses actual RNG to prevent starvation.

**Comment by acw** [comment]
Interesting. Are you planning to fold an RNG into the underlying Chan? I'm assuming that this doesn't need cryptographically-secure randomness, so I suppose you could just use something from `base`?

**Comment by cstrahan** [comment]
> Those look mostly reasonable to me. I'm a little torn, actually. I like this interface, and so integrating something like this into BoundedChan doesn't seem totally out of line. But I do like chanRead throwing an exception sometimes, rather than returning Result. Hmmm. If you're interested, maybe we could include your version, and then create an exception layer on top, too?

You're suggesting that you'd like reading from a closed channel to throw? I suppose you could layer that on top of what I have. I wanted to stick as close as possible to the Go channel semantics, where closing a channel (often a non-buffered channel consisting of zero-width structs, i.e. `make(<-chan struct{}, 0)`) is used for coordination, timeouts being one example (it follows that this approach really only makes sense when the reading of that channel is within a `select`, providing a sort of short-circuit).

Speaking practically, there's a whole lot more bookkeeping due to supporting `select`, and I'm not sure if you'd want to take on that performance hit. If the perf impact is deemed reasonable, there's the API to consider (I'd like to have an API that feels nice and consistent); might it make sense for me to release a separate package that BoundedChan could then layer on its own semantics?

> The one quibble I have is with your Async versions. By the type signatures, they sound like try rather than async. I.e., either this will work or will it fail and nothing will happen. As opposed to async, which I interpret as "this will complete, at some point, probably later." They're both useful, but have different types, a la:

Good point - "try" makes a lot more sense than "async" in those function names.

> Interesting. Are you planning to fold an RNG into the underlying Chan? I'm assuming that this doesn't need cryptographically-secure randomness, so I suppose you could just use something from base?

For the time being, I'm using a simple Fisher-Yates shuffle using `randomRIO`:

``` haskell
shuffleVector
    :: (VGM.MVector v e)
    => v RealWorld e -> IO ()
shuffleVector !xs = do
    let !size = VGM.length xs
    forM_ [1 .. size - 1] $
        \i -> do
            j <- randomRIO (0, i)
            vi <- VGM.read xs i
            vj <- VGM.read xs j
            VGM.write xs j vi
            VGM.write xs i vj
```

All of this has been mostly the product of one 3 day weekend, so I still have some work to do:
- Finish up the naming changes and docs
- Chase down unnecessary allocations (help here would be greatly appreciated)
- Optimizations (ditto)
- Make it all async-exception safe

I probably won't have time to revisit the code until the weekend.
